### PR TITLE
feat(preview1): implement core I/O functionality

### PR DESCRIPTION
**Pull request by rvolosatovs**
Moved from https://github.com/bytecodealliance/preview2-prototyping/pull/175

Refs https://github.com/bytecodealliance/wasmtime/issues/6370

The implementation tries to follow existing preview1 behavior as much as possible - as a result, there are a few differences in behavior from existing adapter. Please see the original PR for more details.

A follow-up issues/PRs fill be filed for:
- `fd_readdir`
- fixing `windows` support for a few tests (still not sure what's causing the failures)
- making all preview1 implementations consistent
- adding rights handling

### CODE REVIEW

**Comment by rvolosatovs** [comment]
cc @pchickey

File: crates/wasi/src/preview2/preview1/mod.rs [hunk 1]
```diff
@@ -1,32 +1,289 @@
...
+    view: &'a mut T,
+    descriptors: Cell<Descriptors>,
+}
+
+impl<T: WasiPreview1View + ?Sized> Drop for DescriptorsRef<'_, T> {
+    fn drop(&mut self) {
+        let descriptors = self.descriptors.take();
+        self.view.adapter_mut().descriptors = Some(descriptors);
+    }
+}
```
Comment (pchickey) [inline crates/wasi/src/preview2/preview1/mod.rs:163]: I wasn't expecting this sort of machinery to be required - we can't get away with implementing the DescriptorsRef methods on WasiPreview1Ctx? I can't intuit where the problem is in a more straightforward implementation, so if this is required, lets leave a comment on the struct definition explaining why.

File: crates/wasi/src/preview2/preview1/mod.rs [hunk 2]
```diff
@@ -253,143 +743,599 @@ impl<
...
         &mut self,
-        fd: types::Fd,
-        fs_rights_base: types::Rights,
-        fs_rights_inheriting: types::Rights,
+        _fd: types::Fd,
+        _fs_rights_base: types::Rights,
+        _fs_rights_inheriting: types::Rights,
     ) -> Result<(), types::Error> {
-        todo!()
+        unreachable!()
```
Comment (pchickey) [inline crates/wasi/src/preview2/preview1/mod.rs:914]: this is still a `todo!()` - we can leave it to a follow up PR, but `unreachable!()` communicates to me that it is a bug in this code if it is reachable, while this is reachable from any guest program.

File: crates/wasi/src/preview2/preview1/mod.rs [hunk 3]
```diff
@@ -253,143 +743,599 @@ impl<
...
+        let path_max_len = path_max_len.try_into().or(Err(types::Errno::Overflow))?;
+        if let Some(Descriptor::PreopenDirectory((_, p))) =
+            self.descriptors().await?.get_descriptor(fd)
+        {
+            // NOTE: This conditional is not valid, the bug is preserved from legacy implementation
+            // for backwards-compatibility
+            // https://github.com/bytecodealliance/wasmtime/blob/b9e4474f1f04ae15ca27ca0ed695d798e1c0a295/crates/wasi-common/src/snapshots/preview_1.rs#L507-L514
+            if p.len() < path_max_len {
+                return Err(types::Errno::Nametoolong.into());
+            }
```
Comment (pchickey) [inline crates/wasi/src/preview2/preview1/mod.rs:1262]: lets definitely not preserve that bug :). That one has been around for a while! Lets get rid of this check and follow up with a PR to fix the legacy impl, and a test. I expect that fix will only be around for its final release, but its still worth doing.

**Review by pchickey — COMMENTED** [review]
This is great work! Just a few small issues. Thank you for finding all of these bad behaviors in the legacy implementation!

File: crates/wasi/src/preview2/preview1/mod.rs (hunk 2 above)
Comment (rvolosatovs) [reply crates/wasi/src/preview2/preview1/mod.rs:914]: I took this directly from https://github.com/bytecodealliance/wasmtime/blob/5b93cdb84def5fa7cb3ff823bc80216f19c296c5/crates/wasi-preview1-component-adapter/src/lib.rs#L560-L569 and it was surprising to me as well.
It seems to me that the adapter has to implement rights handling, which the existing one does not do. I'll mark this as `todo!` and file an issue for rights handling follow-up

File: crates/wasi/src/preview2/preview1/mod.rs (hunk 1 above)
Comment (rvolosatovs) [reply crates/wasi/src/preview2/preview1/mod.rs:163]: I'll add a comment, but the biggest reasons why this is required are:
1. By far the biggest issue is `bindgen` generating host methods with `&mut self` receivers, preventing e.g. simultaneous lazy initialization of the descriptor table and calling a preview2 method using any data borrowed from that table, even looking up something like `table().is_file` is problematic
2. `WasiPreview1Adapter` (assuming that's what you meant by WasiPreview1Ctx) does not have access to preview2 functionality and/or the underlying table, so it cannot initialize itself using something like https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html . Lazily initializing the descriptor map requires access to both the `WasiPreview1Adapter` and `wasi::preopens::Host`, so the only place where this functionality could exist is in an extension trait (`WasiPreview1ViewExt`). Note that from within this trait, we can never return anything borrowed directly, because we'd first need to call `adapter_mut()` to get access to the state (which we may need to lazily initialize). And even if we had a way to return a borrowed descriptor table directly, it would not be possible to call any other method on `self` due to 1., since that would cause a double mutable borrow, which is not allowed.

In short, an alternative solution would be storing `Option<Arc<Mutex<Descriptors>>>` in the adapter and returning `Arc<Mutex<Descriptors>>` in the `WasiPreview1AdapterExt::descriptors(&mut self)` method, which IMO is way more awkward to work with and simply redundant complexity.
Overall, the `DescriptorsRef` is just a workaround for `bindgen` design drawbacks. I believe this could be simplified dramatically if host methods would just take `&self`

File: crates/wasi/src/preview2/preview1/mod.rs (hunk 2 above)
Comment (pchickey) [reply crates/wasi/src/preview2/preview1/mod.rs]: Oh, that is indeed a bug in the component adapter as well. `unreachabable!()` is just how we spell `todo!()` and `panic!()` and etc over there since we cant use std's macros.

In the legacy implementation, set_rights is a no-op, it only validates the fd. That is the correct behavior for both of these spots. Would you mind making this fix to the component adapter as well?

File: crates/wasi/src/preview2/preview1/mod.rs (hunk 1 above)
Comment (pchickey) [reply crates/wasi/src/preview2/preview1/mod.rs:180]: That makes sense. Thank you.

We should look into adding a bindgen option to enable &self receivers in the future, and then we may be able to unwind a bit of this complexity. For now, this is a fine implementation once it has a comment.

Comment (rvolosatovs) [reply crates/wasi/src/preview2/preview1/mod.rs:180]: I've added a few comments and renamed the struct to `Transaction` so it's a bit more obvious what it actually is

File: crates/wasi/src/preview2/preview1/mod.rs (hunk 2 above)
Comment (rvolosatovs) [reply crates/wasi/src/preview2/preview1/mod.rs]: Let's take care of it as part of https://github.com/bytecodealliance/wasmtime/issues/6446 ?

**Review by pchickey — APPROVED** [review]
